[ELY-2496] Add integrity to existing filesystem realms using Elytron Tool

jessicarod7 opened the pull request:
https://issues.redhat.com/browse/ELY-2496

**Commits:**

~~1. The first two commits come from ELY-2514, since the integrity command uses JDK 11 features. First JDK 11 commit removes Java 8 specific classes~~
~~2. Second JDK 11 commit replaces with Java 9 specific classes  and updates POMs.~~ **(dropped from this PR)**
1. Moves reused parameters into a common class [Params](https://github.com/wildfly-security/wildfly-elytron/blob/bfa9d7a19297e5ef8ae7f26353579f9501d852b3/tool/src/main/java/org/wildfly/security/tool/Command.java#L364-L398) within Command 
2. Add `hash-charset` parameter and test cases to _filesystem-realm-encrypt_, which was missed in the original PR #1676 
3. **EDIT:** Updated warning messages in `Command.getSecretKey()` to be more consistent, and always warn (vs. sometimes crashing from an uncaught exception)
4. Adds new _filesystem-realm-integrity_ command, with support for upgraded encrypted realms
5. Support upgrading realms with integrity enabled in _filesystem-realm-encrypt_. Also disabled upgrade of empty realms (see below)
6. Add test cases in both commands for filesystem realms with integrity
7. Modified to remove methods which require JDK 11 (see #1880)

**Summary of changes:**

- Addition of a new command _filesystem-realm-integrity_ to upgrade existing realms to support integrity verification. Requires an non-empty filesystem realm, and a password-protected keystore with a key pair.
  - Supports upgrading encrypted realms
  - Supports in-place realm upgrade (upgraded realm is placed in the same directory)
  - Test cases for all params, with validation of CLI scripts via [`validateScript` method and `ScriptParams` class](https://github.com/wildfly-security/wildfly-elytron/blob/95b1004a91d6ca91172ac9a9819654ceb4746e16/tool/src/test/java/org/wildfly/security/tool/FileSystemRealmIntegrityCommandTest.java#L588)
  - Encryption command issues not present in integrity command:  [ELY-2530](https://issues.redhat.com/browse/ELY-2530), [ELY-2536](https://issues.redhat.com/browse/ELY-2536), [ELY-2540](https://issues.redhat.com/browse/ELY-2540), [ELY-2541](https://issues.redhat.com/browse/ELY-2541), [ELY-2542](https://issues.redhat.com/browse/ELY-2542), and [ELY-2546](https://issues.redhat.com/browse/ELY-2546)
- Update _filesystem-realm-encrypt_ command to support upgrading filesystem realms which already have integrity enabled (Note that encryption command cannot enable integrity, and vice versa)
  - Added missing option `hash-charset` to encryption command
- Changes for both commands
  - Move static Elytron password/keystore providers, overrideable warning/error handlers, and methods `getSecretKey` and `getKeyPair` to `Command` class
  - Add new method `FileSystemRealmUtil.createRealmWithIntegrityValidation()` to build new realm; moved shared code between encryption and integrity commands into a private method
  - Empty filesystem realms cannot upgraded by Elytron Tool (the directory is not copied), so instead provides a warning and instructs user to update within management client (see wildfly/wildfly-core#5411)
- Other
  - Move common parameters and strings across Elytron Tool commands (ex. `HELP_PARAM`, `System.getProperty(“line.separator”)`) to new class `org.wildfly.security.tool.Command$Params`

Also, I've manually imported and tested all of the test case realms with the wildfly/wildfly@784cc50 commit. All identities are readable, and the `:verify-integrity` CLI command works.

jessicarod7 committed:
[ELY-2496] Move common methods and variables to Command class

jessicarod7 committed:
[ELY-2496] Add missing hash-charset option to FileSystemEncryptRealmCommand

jessicarod7 committed:
[ELY-2496] Add integrity to existing filesystem realms using Elytron Tool

jessicarod7 committed:
[ELY-2496] FileSystemEncryptRealmCommand supports realms with integrity

jessicarod7 committed:
[ELY-2496] Add integrity command test cases

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -204,6 +221,28 @@ public void printDuplicatesWarning(CommandLine cmdLine, List<String> duplicatesA
 
+    /**
+     * Prints out a warning message if silentMode is not enabled and adds the warning to the summary
+     * if summaryMode is enabled
```
Just to check, the logic for determining if silentMode is not enabled and determining if summaryMode is enabled is handled elsewhere right. (This is just a minor comment, just that the description of the method doesn't match what the method is doing.)

fjuma replied:
Ah, I see now that these methods aren't new. They were just moved.

jessicarod7 replied:
Yeah, I missed these comments. I'll update them both

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -204,6 +221,28 @@ public void printDuplicatesWarning(CommandLine cmdLine, List<String> duplicatesA
+    }
+
+    /**
+     * Determines if a summary needs to be printed and prints summary after an error is thrown
```
Similar comment here, this method doesn't determine if the summary needs to be printed, right. It looks like it just sets the status.

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +309,91 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+        SecretKey key;
+        try {
+            key = credentialStore.retrieve(alias, SecretKeyCredential.class).getSecretKey();
+        } catch (NullPointerException e) {
```
Instead of catching the `NullPointerException` here, it would be better to check the return value of `credentialStore.retrieve` prior to calling `getSecretKey`. If it's `null`, we can print the error and return `null`.

jessicarod7 replied:
I'll update this and the other section to be similar to `getKeyPair`. This code was unchanged from the original encrypt command, so I'm not as familiar with the functionality.

jessicarod7 replied:
@OndrejKotek Noting that I made some changes in response to this, in the new 5th commit

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +309,91 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+        implProps.put("modifiable", Boolean.TRUE.toString());
+        credentialStore.initialize(implProps);
+        try {
+            credentialStore.retrieve(alias, SecretKeyCredential.class).getSecretKey();
```
The return value should be checked before calling `getSecretKey`.

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemRealmUtil.java:
```diff
@@ -42,23 +44,38 @@ public class FileSystemRealmUtil {
+     */
+    public static void createRealmWithIntegrityValidation(FileSystemSecurityRealm inputFileSystemRealm,
+                                  FileSystemSecurityRealm outputFileSystemRealm) throws RealmUnavailableException {
+        transferIdentitiesToNewRealm(inputFileSystemRealm, outputFileSystemRealm);
```
What is this method intended to do? It looks like both this method and `createEncryptedRealmFromUnencrypted` currently do the same thing which is just call `transferIdentitiesToNewRealm`. Perhaps the name / description of these methods just needs to be updated? The current description here says the outputFileSystemRealm is also configured with a key pair to verify the integrity but it doesn't look like there's anything specific to integrity validation here.

jessicarod7 replied:
Missed this, it was a leftover from when I wasn't sure if they would need separate functionality

jessicarod7 committed:
[ELY-2496] Update warnings for SecretKey retrieval

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +309,91 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+    SecretKey getSecretKey(Boolean createCredentialStore, String credentialStoreLocation, String alias, Boolean populate,
+                           String inputRealmLocation) throws Exception {
+        CredentialStore credentialStore;
+        // check if credential-store and secret-key-alias are both specified, or both null
```
What code is this comment referring to? I don't see checks for the credential store location or alias here.

jessicarod7 replied:
looks like a leftover from when the code was in the encryption class, removed

fjuma commented on auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemRealmUtil.java:
```diff
@@ -31,34 +32,34 @@
-    public static void createEncryptedRealmFromUnencrypted(FileSystemSecurityRealm unencryptedRealm, FileSystemSecurityRealm encryptedRealm) throws RealmUnavailableException {
-        Assert.checkNotNullParam("unencryptedRealm", unencryptedRealm);
-        Assert.checkNotNullParam("encryptedRealm", encryptedRealm);
+    public static void cloneIdentitiesToNewRealm(FileSystemSecurityRealm oldRealm, FileSystemSecurityRealm newRealm) throws RealmUnavailableException {
```
FYI for other reviewers, this is private API.

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +326,177 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+     * @return the requested {@link SecretKey}, or {@code null} if it could not be retrieved
+     */
+    SecretKey getSecretKey(Boolean createCredentialStore, String credentialStoreLocation, String alias, Boolean populate,
+                           String inputRealmLocation, int descriptorBlockCount) throws Exception {
```
Does this throw anything other than `CredentialStoreException`? Can we list the specific exceptions?

fjuma replied:
Have the parameters been validated before this helper method is called?

jessicarod7 replied:
> Does this throw anything other than `CredentialStoreException`? Can we list the specific exceptions?

Sure, the other exception is `NoSuchAlgorithmException` from getting a cred store instance

> Have the parameters been validated before this helper method is called?

They should at least be non-null, but looks like the credential store isn't checked to see if the path exists. I can update that in both commands

fjuma replied:
If we've already checked for non-null before calling this helper method, that's fine. I don't think we need to validate the path exists at this point. I believe that will be checked when calling credentialStore.initialize below.

jessicarod7 replied:
alright, I'll leave it as is

jessicarod7 replied:
Double checked, `Exception` has to be thrown because `errorHandler` is called, so I'll specify in the javadoc

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +326,177 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+     * @return the requested {@link KeyPair}, or {@code null} if it could not be retrieved
+     */
+    KeyPair getKeyPair(Path keyStorePath, String keyStoreType, String keyPairAlias, char[] password, int descriptorBlockCount) throws Exception {
+        AtomicLoadKeyStore keyStore = null;
```
Just to make sure, have the parameters been validated before this helper method is called? (e.g., keyPairAlias will be non-null, etc.)

jessicarod7 replied:
will validate key store path in encryption command, but otherwise yes, everything exists and paths should exist

jessicarod7 replied:
just replied to the other comment, I'll still update this one so it's consistent with the call from the integrity command, but otherwise it should be fine as is

fjuma commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -270,4 +326,177 @@ protected Supplier<Provider[]> getProvidersSupplier(final String providersList)
+     * @param descriptorBlockCount index of a descriptor block from bulk conversion files
+     * @return the requested {@link KeyPair}, or {@code null} if it could not be retrieved
+     */
+    KeyPair getKeyPair(Path keyStorePath, String keyStoreType, String keyPairAlias, char[] password, int descriptorBlockCount) throws Exception {
```
Which exceptions can be thrown here?

jessicarod7 replied:
updated

fjuma commented on tool/src/main/java/org/wildfly/security/tool/CredentialStoreCommand.java:
```diff
@@ -50,12 +64,11 @@
 import org.wildfly.common.iteration.CodePointIterator;
-import org.wildfly.security.auth.util.ElytronFilePasswordProvider;
 import org.wildfly.security.auth.server.IdentityCredentials;
+import org.wildfly.security.auth.util.ElytronFilePasswordProvider;
```
It's good to avoid reordering imports since this type of change isn't related to the PR.

jessicarod7 replied:
ah, eclipse formatter kicking in, will double check in all commits

fjuma commented on tool/src/main/java/org/wildfly/security/tool/ElytronTool.java:
```diff
@@ -17,14 +17,14 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import org.apache.commons.cli.AlreadySelectedException;
```
Same note here about import reordering.

jessicarod7 committed:
[ELY-2496] Modify commands to support JDK 8

Due to ELY-2514, this module must be compiled for Java 8. This commit
can be reversed or reverted once this issue is fixed.

jessicarod7 commented:
Looks like a random failure due to a [timeout on the Windows build](https://github.com/wildfly-security/wildfly-elytron/actions/runs/5024942363/jobs/9011265598?pr=1896#step:6:4156). Could someone please rerun the CI?

fjuma commented:
Thanks, re-triggered it.

skyllarr commented on tool/src/main/java/org/wildfly/security/tool/Command.java:
```diff
@@ -204,6 +241,27 @@ public void printDuplicatesWarning(CommandLine cmdLine, List<String> duplicatesA
+     * @param warning The warning to be shown
+     */
+    protected void warningHandler(String warning) {
+        System.out.print("WARNING: ");
```
@cam-rod just a minor,  we can set status to `GENERAL_CONFIGURATION_WARNING` here

jessicarod7 replied:
~~fixed~~

jessicarod7 replied:
turns out this change actually broke a number of tests in FileSystemRealmCommandTest, so I'm reverting it

skyllarr commented on tool/src/main/java/org/wildfly/security/tool/FileSystemRealmIntegrityCommand.java:
```diff
@@ -0,0 +1,1073 @@
+            if (passwordOption == null && passwordEnvOption == null) {
+                passwordOption = prompt(false, ElytronToolMessages.msg.keyStorePasswordPrompt(), false, null);
+                if (passwordOption == null) {
+                    setStatus(GENERAL_CONFIGURATION_ERROR);
```
@cam-rod just a total minor, this line is superfluous since we configured the status to be `GENERAL_CONFIGURATION_ERROR` at the beginning

jessicarod7 replied:
fixed in this command and encryption (where I made the same change)

skyllarr reviewed (approved):
@cam-rod LGTM, thanks!

fjuma reviewed (approved):
Thanks @cam-rod!

fjuma commented:
@cam-rod Just noticed, we have some CI failures now.

jessicarod7 commented:
Thanks for the heads up, added some patches for the rebase.

fjuma commented:
Thanks @cam-rod!
